### PR TITLE
[SPARK-23518][SQL] Avoid metastore access when the users only want to read and write data frames

### PR DESCRIPTION
**Pull request by liufengdb**
## What changes were proposed in this pull request?

https://github.com/apache/spark/pull/18944 added one patch, which allowed a spark session to be created when the hive metastore server is down. However, it did not allow running any commands with the spark session. This brings troubles to the user who only wants to read / write data frames without metastore setup.

## How was this patch tested?

Added some unit tests to read and write data frames based on the original HiveMetastoreLazyInitializationSuite.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by liufengdb** [comment]
@felixcheung Can you take a look at the changes in the R tests?

**Review by gatorsmile — COMMENTED** [review]
LGTM except R-related codes. cc @felixcheung

File: R/pkg/tests/fulltests/test_sparkSQL.R [hunk 1]
```diff
@@ -46,16 +49,19 @@ setHiveContext <- function(sc) {
...
   hiveSession
 }
 
 unsetHiveContext <- function() {
   previousSession <- get(".prevSparkRsession", envir = .sparkREnv)
+  previousConf <- get(".prevSessionConf", envir = .sparkREnv)
+  callJStatic("org.apache.spark.sql.api.r.SQLUtils", "setSparkContextSessionConf", previousSessioni, previousConf)
   assign(".sparkRsession", previousSession, envir = .sparkREnv)
   remove(".prevSparkRsession", envir = .sparkREnv)
+  remove(".prevSparkConf", envir = .sparkREnv)
```
Comment (felixcheung) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: this should be `. prevSessionConf`?

File: R/pkg/tests/fulltests/test_sparkSQL.R [hunk 2]
```diff
@@ -46,16 +49,19 @@ setHiveContext <- function(sc) {
...
   assign(".sparkRsession", hiveSession, envir = .sparkREnv)
   assign(".prevSparkRsession", previousSession, envir = .sparkREnv)
+  assign(".prevSessionConf", previousConf, envir = .sparkREnv)
   hiveSession
 }
 
 unsetHiveContext <- function() {
   previousSession <- get(".prevSparkRsession", envir = .sparkREnv)
+  previousConf <- get(".prevSessionConf", envir = .sparkREnv)
+  callJStatic("org.apache.spark.sql.api.r.SQLUtils", "setSparkContextSessionConf", previousSessioni, previousConf)
```
Comment (felixcheung) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: this should be `previousSession`?

**Comment by felixcheung** [comment]
thx, let's wait for tests

File: R/pkg/tests/fulltests/test_sparkSQL.R [hunk 3]
```diff
@@ -885,24 +894,24 @@ test_that("collect() and take() on a DataFrame return the same number of rows an
...
 
 test_that("collect() support Unicode characters", {
-  lines <- c("{\"name\":\"안녕하세요\"}",
-             "{\"name\":\"您好\", \"age\":30}",
-             "{\"name\":\"こんにちは\", \"age\":19}",
-             "{\"name\":\"Xin chào\"}")
+  lines <- c("{\"name\":\"?????\"}",
+             "{\"name\":\"??\", \"age\":30}",
+             "{\"name\":\"?????\", \"age\":19}",
+             "{\"name\":\"Xin ch?o\"}")
```
Comment (ueshin) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: Is it okay to replace these unicode characters?

Comment (felixcheung) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: no, we should not change these

**Review by felixcheung — COMMENTED** [review]
we need to make a few changes

**Review by felixcheung — APPROVED** [review]
LG this should do it...

**Comment by gatorsmile** [comment]
retest this please

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by liufengdb** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by felixcheung** [comment]
looks like test failures are related?

**Comment by liufengdb** [comment]
My original plan to fix the test should not work, because of this test: https://github.com/apache/spark/blob/master/R/pkg/tests/fulltests/test_sparkSQL.R#L3343

The new plan is to run some simple catalog commands immediately after the spark session is created, so the catalog is materialized (like the old behavior).

**Comment by gatorsmile** [comment]
Can we fix this test? https://github.com/apache/spark/blob/master/R/pkg/tests/fulltests/test_sparkSQL.R#L3343

**Comment by liufengdb** [comment]
We can remove the test, but it is not a good practice. You don't know exactly why the test is added, which hidden assuption he wants to guarantee, right?

**Comment by liufengdb** [comment]
Overall, I think this suite needs a refactoring: split to in-memory catalog one and hive catalog one. The catalog conf should not be manipulated after the spark context is created. The other way is just a hack.

**Comment by gatorsmile** [comment]
cc @felixcheung Does this PR look good to you now?

**Comment by liufengdb** [comment]
retest this please

File: R/pkg/tests/fulltests/test_sparkSQL.R [hunk 4]
```diff
@@ -67,6 +67,8 @@ sparkSession <- if (windows_with_hadoop()) {
     sparkR.session(master = sparkRTestMaster, enableHiveSupport = FALSE)
   }
 sc <- callJStatic("org.apache.spark.sql.api.r.SQLUtils", "getJavaSparkContext", sparkSession)
+# materialize the catalog implementation
+listTables()
```
Comment (felixcheung) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: we are calling `sparkR.session(master = sparkRTestMaster, enableHiveSupport = FALSE)` is almost every other test files - does the same apply in those other places?

**Review by felixcheung — COMMENTED** [review]
I think this test here https://github.com/apache/spark/blob/master/R/pkg/tests/fulltests/test_sparkSQL.R#L3343 is just to see if the spark.sql.catalogImplementation is set, it could easier be done in a different way if that is an issue.

File: R/pkg/tests/fulltests/test_sparkSQL.R (hunk 4 above)
Comment (liufengdb) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: `test_sparkSQL.R` is the only one uses `newJObject("org.apache.spark.sql.hive.test.TestHiveContext", ssc, FALSE)` on the `ssc`, so the catalog impl spark conf is changed. So ``test_sparkSQL.R` is the only one broken.

Comment (felixcheung) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: ok

**Comment by gatorsmile** [comment]
LGTM 

Thanks! Merged to master.

**Pull request closed without merging**